protocol/communities_messenger_test.go

```diff
@@ -3684,3 +3684,115 @@ func (s *MessengerCommunitiesSuite) TestRetrieveBigCommunity() {
+	} else {
+		s.Require().Equal(allCommunities[1].RequestedToJoinAt(), requestToJoin1.Clock)
+	}
+
+	// pull to make sure it has been saved
+	requestsToJoin, err := s.alice.MyPendingRequestsToJoin()
+	s.Require().NoError(err)
+	s.Require().Len(requestsToJoin, 1)
+
+	// Make sure the requests are fetched also by community
+	requestsToJoin, err = s.alice.PendingRequestsToJoinForCommunity(community.ID())
+	s.Require().NoError(err)
+	s.Require().Len(requestsToJoin, 1)
+
+	requestToJoin2 := response.RequestsToJoinCommunity[0]
+
+	s.Require().NotNil(requestToJoin2)
+	s.Require().Equal(community.ID(), requestToJoin2.CommunityID)
+	s.Require().NotEmpty(requestToJoin2.ID)
+	s.Require().NotEmpty(requestToJoin2.Clock)
+	s.Require().Equal(requestToJoin2.PublicKey, common.PubkeyToHex(&s.alice.identity.PublicKey))
+	s.Require().Equal(communities.RequestToJoinStatePending, requestToJoin2.State)
+
+	s.Require().Equal(requestToJoin1.ID, requestToJoin2.ID)
+
+	requestToCancel := &requests.CancelRequestToJoinCommunity{ID: requestToJoin1.ID}
+	response, err = s.alice.CancelRequestToJoinCommunity(ctx, requestToCancel)
+	s.Require().NoError(err)
+	s.Require().NotNil(response)
+	s.Require().Len(response.RequestsToJoinCommunity, 1)
+	s.Require().Equal(communities.RequestToJoinStateCanceled, response.RequestsToJoinCommunity[0].State)
+
+	// pull to make sure it has been saved
+	cancelRequestsToJoin, err := s.alice.MyCanceledRequestToJoinForCommunityID(community.ID())
+	s.Require().NoError(err)
+	s.Require().NotNil(cancelRequestsToJoin)
+	s.Require().Equal(cancelRequestsToJoin.State, communities.RequestToJoinStateCanceled)
+
+	response, err = WaitOnMessengerResponse(
+		s.owner,
+		func(r *MessengerResponse) bool { return len(r.RequestsToJoinCommunity) >= 1 },
+		"no requests to join community",
+	)
+
+	s.Require().NoError(err)
+
+	_, err = WaitOnMessengerResponse(
+		s.alice,
+		func(r *MessengerResponse) bool { return r != nil },
+		"no communities",
+	)
+
+	s.Require().NoError(err)
+	// Retrieve activity center notifications for admin to make sure the request notification is deleted
+	notifications, err := s.owner.ActivityCenterNotifications(ActivityCenterNotificationsRequest{
+		Cursor:        "",
+		Limit:         10,
+		ActivityTypes: []ActivityCenterType{},
+		ReadType:      ActivityCenterQueryParamsReadUnread,
+	})
+
+	s.Require().NoError(err)
+	s.Require().Len(notifications.Notifications, 0)
+
+	cancelRequestToJoin2 := response.RequestsToJoinCommunity[0]
+	s.Require().NotNil(cancelRequestToJoin2)
+	s.Require().Equal(community.ID(), cancelRequestToJoin2.CommunityID)
+	s.Require().False(cancelRequestToJoin2.Our)
+	s.Require().NotEmpty(cancelRequestToJoin2.ID)
+	s.Require().NotEmpty(cancelRequestToJoin2.Clock)
+	s.Require().Equal(cancelRequestToJoin2.PublicKey, common.PubkeyToHex(&s.alice.identity.PublicKey))
+	if cancelRequestsToJoin.Clock > requestToJoin1.Clock {
+		s.Require().False(response.Communities()[0].HasMember(&s.alice.identity.PublicKey))
+	}
```

Comment:
It seems this condition makes the test not deterministic? Sometimes it checks what we are interested in (alice not being a member of community), the other time it is not.

Reply:
I thought it ensured correct message order.
Thinking of a workaround.